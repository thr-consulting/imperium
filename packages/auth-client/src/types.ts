export interface ImperiumAuthClientModule {
	thing?: string;
}

export interface LoginInfo {
	identifier: string;
	password: {
		digest: string;
		algorithm: string;
	};
	rememberDevice?: boolean;
}

export interface LoginReturn {
	id: string;
	access: string;
}

export interface AccessToken {
	id: string;
	roles?: string[];
	iat: number;
	exp: number;
}
