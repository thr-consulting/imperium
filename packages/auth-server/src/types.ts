/* eslint-disable import/no-cycle */
// see: https://github.com/babel/babel/issues/10981
import {IContextManager} from '@imperium/server';
import {AuthModuleContext} from './AuthServerModule';

export interface LoginInfo {
	identifier: string;
	password: {
		digest: string;
		algorithm: string;
	};
}

export function isLoginInfo(loginInfo: object): loginInfo is LoginInfo {
	return (
		(loginInfo as LoginInfo).identifier !== undefined &&
		(loginInfo as LoginInfo).password !== undefined &&
		(loginInfo as LoginInfo).password.algorithm !== undefined &&
		(loginInfo as LoginInfo).password.digest !== undefined
	);
}

export interface LoginReturn {
	id: string;
	access: string;
	refresh: string;
}

export interface ServiceInfo {
	id: string;
	roles: string[];
	password: {
		bcrypt: string;
	};
	blacklist?: number[];
}

export interface RefreshToken {
	id: string;
	type: string;
	iat: number;
	exp: number;
}

export function isRefreshToken(refreshToken: object): refreshToken is RefreshToken {
	return (
		(refreshToken as RefreshToken).id !== undefined && (refreshToken as RefreshToken).exp !== undefined && (refreshToken as RefreshToken).type === 'r'
	);
}

export interface AccessToken {
	id: string;
	roles?: string[];
	iat: string;
	exp: string;
}

export function isAccessToken(accessToken: object): accessToken is AccessToken {
	return (
		(accessToken as AccessToken).id !== undefined && (accessToken as AccessToken).iat !== undefined && (accessToken as AccessToken).exp !== undefined
	);
}

export type AuthContextManager = IContextManager<ReturnType<typeof AuthModuleContext>>;

export interface ImperiumAuthServerModule<ContextManager extends AuthContextManager = AuthContextManager> {
	auth?: {
		getServiceInfo: (identifier: string, ctx: ContextManager) => Promise<ServiceInfo | null>;
	};
}
