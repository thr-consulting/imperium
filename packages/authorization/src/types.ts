import type {Authorization} from './Authorization';

export interface IndexedValue<Value> {
	index: number;
	value: Value;
}

export type JsonValue = string | number | boolean | null | JsonValue[] | {[key: string]: JsonValue};

/**
 * A Permission is either a string or a string and Json data.
 */
export type Permission = string;
export interface PermissionWithData {
	permission: string;
	data?: JsonValue;
}
export type Permissions = Permission | PermissionWithData | (Permission | PermissionWithData)[];

export type WithKey<T extends object> = T & {id: string | null};

export interface PermissionLookupOpts<Extra extends AuthenticationBase> {
	keys: readonly WithKey<PermissionWithData>[];
	authorization: Authorization<Extra>;
}

export type PermissionLookup<Extra extends AuthenticationBase> = (opts: PermissionLookupOpts<Extra>) => Promise<boolean[]>;

/**
 * A simple cache interface (used for client side caching)
 */
export interface AuthorizationCache {
	get(key: string): Promise<any>;
	set(key: string, data: any, expire?: number): Promise<any>;
	exists(key: string): Promise<boolean>;
}

export interface AuthenticationBase {
	auth?: {
		id?: string;
	};
}

export interface AuthenticationRequest extends AuthenticationBase {
	hostname?: string;
	ip?: string;
	headers?: {
		[key: string]: string;
	};
}

export interface AuthenticationToken extends AuthenticationBase {
	getToken: () => Promise<string | null>;
}
