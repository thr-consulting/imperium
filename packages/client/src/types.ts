import type React from 'react';

export type ImperiumEnvironmentVar = string | number | boolean | ImperiumEnvironment;
export type ImperiumEnvironment = {[key: string]: ImperiumEnvironmentVar | ImperiumEnvironmentVar[]};

export type GlobalConst = {[key: string]: string | number | boolean};
export type RootProps = {[key: string]: any};
export type Hoc = (
	WrappedComponent: React.ComponentType,
) => {
	(props: any): React.ReactNode;
	displayName: string;
};
export type HocCreator = (client: IImperiumClient) => Hoc;

export interface ImperiumClientModule {
	name: string;
	environment?: (globalConst: GlobalConst, currentEnvironment: ImperiumEnvironment) => Promise<ImperiumEnvironment>;
	startup?: (client: IImperiumClient) => RootProps | void;
	hocs?: HocCreator[];
}

export type ImperiumClientModuleFunction = () => ImperiumClientModule;

export interface IImperiumClient {
	start(): Promise<void>;
	readonly modules: ImperiumClientModule[];
	readonly environment: ImperiumEnvironment;
	readonly globalConst: GlobalConst;
}
