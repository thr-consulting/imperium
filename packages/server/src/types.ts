import type {Connector} from '@imperium/context-manager';
import type {ImperiumServer} from './ImperiumServer';

/**
 * Used to define custom server modules.
 */
export interface ImperiumServerModule<Context, Connectors extends Connector> {
	name: string;
	endpoints?: (server: ImperiumServer<Context, Connectors>) => void;
	startup?: (server: ImperiumServer<Context, Connectors>, context: Context) => Promise<void>;
}

/**
 * The config required to create a new ImperiumServer.
 */
export interface ImperiumServerConfig<Context, Connectors extends Connector> {
	connectors: Connectors;
	serverModules: () => ImperiumServerModule<Context, Connectors>[];
	contextCreator: (connector: Connectors) => Context; // TODO add auth here
}

/**
 * The shape of the configuration object used to configure Imperium Dev launcher.
 */
export interface ImperiumConfig {
	development?: {
		clientPort?: number;
		workerCrashDelay?: number;
		workerCrashMax?: number;
		imperiumDevelopmentAliases?: boolean;
	};
	production?: {
		path?: string;
		client?: {
			minimize?: boolean;
			devtool?: boolean;
			vendorChunk?: string[];
		};
		server?: {
			minimize?: boolean;
			devtool?: boolean;
			externals?: string[];
		};
	};
	webpack?: {
		client?: {
			rules: any[];
		};
		server?: {
			rules: any[];
		};
	};
	source?: {
		projectRoot?: string;
		imperiumRoot?: string;
		path?: string;
		serverIndex?: string;
		clientIndex?: string;
		configModules?: string;
		watchPaths?: string[];
	};
	html?: {
		template?: string;
		meta?: {
			[key: string]: string;
		};
		templateParameters?: {
			[key: string]: string;
		};
	};
}
