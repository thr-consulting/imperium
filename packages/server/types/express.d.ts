import {IContextManager} from '../src';

declare module 'express-serve-static-core' {
	interface Request {
		contextManager: IContextManager;
	}
}
