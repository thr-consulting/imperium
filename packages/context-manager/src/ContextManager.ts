import type {Connector} from './Connector';
import {Auth} from './Auth';

interface ContextCreators<C extends Connector> {
	[key: string]: (conn: C) => any;
}

export class ContextManager<T extends ContextCreators<C>, C extends Connector> {
	public readonly auth: Auth;
	public readonly context = {} as {[P in keyof T]: ReturnType<T[P]>};

	constructor(context: T, connectors: C, auth: Auth = new Auth()) {
		// ContextManager is meant to be created frequently, so the connectors should be connected already.
		if (!connectors.isConnected) {
			throw new Error('Connectors must be connected before creating ContextManager');
		}
		(Object.keys(context) as (keyof T)[]).forEach(key => {
			// creating the contexts
			if (typeof context[key] === 'function') {
				this.context[key] = context[key](connectors);
			}
		});

		this.auth = auth;
	}
}
