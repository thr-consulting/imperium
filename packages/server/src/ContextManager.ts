import {ContextMap, ContextMapFunc, IContextManager, IImperiumServer, Context} from './types';

export class ContextManager implements IContextManager {
	[prop: string]: Context;
	readonly _server: IImperiumServer;
	private _context: ContextMap;
	private _auth: any;

	constructor(server: IImperiumServer) {
		this._server = server;
		this._context = {};
		this._auth = null;
	}

	/**
	 * Adds a module function to the registry. A module function has the
	 * following signature: (connectors, context) => {} and returns
	 * an object keyed with data model objects.
	 * @param contextFunc
	 */
	addContext(contextFunc: ContextMapFunc): void {
		const context = contextFunc(this._server, this);
		const keys = Object.keys(context);
		for (let i = 0; i < keys.length; i++) {
			const key = keys[i];
			if (this[key] === undefined) {
				this[key] = context[key];
			}
		}
	}

	/**
	 * Returns the model with the specified name
	 * @param name
	 * @returns {*}
	 */
	getContext(name: string): Context {
		return this._context[name];
	}

	/**
	 * Returns all the models
	 * @returns {*}
	 */
	get context(): ContextMap {
		return this._context;
	}

	/**
	 * Sets new auth data
	 * @param value
	 */
	set auth(value) {
		this._auth = value;
	}

	/**
	 * Returns the auth data
	 * @returns {null|*}
	 */
	get auth(): any {
		return this._auth;
	}

	get server(): IImperiumServer {
		return this._server;
	}
}
