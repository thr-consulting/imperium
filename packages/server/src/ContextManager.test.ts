import {ContextManager} from './ContextManager';
import {IImperiumServer} from './types';

// @ts-ignore
const mockServer: IImperiumServer = {
	connectors: {
		MyConnector: {},
	},
	environment: {},
	addEnvironment(): void {},
	modules: [],
};

function moduleFunc(/* server */) {
	return {
		MyDataModel: {
			fakeApiFn: () => 'fake api result',
		},
	};
}

describe('Context', () => {
	it('should add modules properly', () => {
		const context = new ContextManager(mockServer);
		context.addContext(moduleFunc);
		expect(context.MyDataModel).toMatchSnapshot();
	});

	it('should set and get auth data', () => {
		const context = new ContextManager(mockServer);
		context.auth = {name: 'test'};
		expect(context.auth).toMatchSnapshot();
	});

	it('should retrieve the connectors', () => {
		const context = new ContextManager(mockServer);
		expect(context.connectors).toMatchSnapshot();
	});
});
