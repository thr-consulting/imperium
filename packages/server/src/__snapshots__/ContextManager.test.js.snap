// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context should retrieve the connectors 1`] = `undefined`;

exports[`Context should set and get auth data 1`] = `
Object {
  "name": "test",
}
`;
